src/Components/Components/src/RenderTree/EventHandlerDelegateWithClosuresComparer.cs

```diff
@@ -0,0 +1,130 @@
+            // the EventCallbacks are not equal.
+            if (left.Receiver != right.Receiver &&
+                (left.RequiresExplicitReceiver || right.RequiresExplicitReceiver))
+            {
+                return false;
+            }
+
+            return DelegateEquals(left.Delegate, right.Delegate);
+        }
+
+        internal static bool DelegateEquals(MulticastDelegate? left, MulticastDelegate? right)
+        {
+            if (Equals(left, right))
+            {
+                return true;
+            }
+
+            // If any of the delegates or their targets are null (static delegates) then normal
+            // equality is sufficient. No additional testing required.
+            if (left == null || right == null || left.Target == null || right.Target == null)
+            {
+                return false;
+            }
+
+            var oldTargetType = left.Target!.GetType();
+            var newTargetType = right.Target!.GetType();
+
+            // if the types are not the same, or the methods are not the same, these possible closures are
+            // not the same anyway.
+            if (oldTargetType != newTargetType || left.Method != right.Method)
+            {
+                return false;
+            }
+
+            // for speed we cache the comparison functions per closure type.
+            var comparison = ClosureComparerCache.GetOrAdd(newTargetType, GetClosureComparerForType);
+            return comparison(left.Target, right.Target);
+        }
+
+        private static Func<object, object, bool> GetClosureComparerForType([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicFields)] Type type)
+        {
+            // if this type is not compiler generated, it is not a closure, and the previous tests where
+            // enough to conclude that these are not equal. We just cache a delegate to a function that returns false.
+            if (!type.CustomAttributes.Any(x => x.AttributeType == typeof(CompilerGeneratedAttribute)))
+            {
+                return (objA, objB) => false;
+            }
+
+            // build an array of lambdas that will get each public field value on the closure.
+            var fieldGetters = type.GetFields(
+                BindingFlags.Public |
+                BindingFlags.Instance |
+                BindingFlags.DeclaredOnly).Select(fieldInfo =>
+            {
+                var sourceParam = Expression.Parameter(typeof(object));
+                Expression returnExpression = Expression.Field(Expression.Convert(sourceParam, fieldInfo.DeclaringType!), fieldInfo);
+                if (fieldInfo.FieldType.IsValueType)
+                {
+                    // box if neccessary.
+                    returnExpression = Expression.Convert(returnExpression, typeof(object));
+                }
+                var lambda = Expression.Lambda(returnExpression, sourceParam);
+                return (Func<object?, object?>)lambda.Compile();
```

Comment:
I'm not sure this is not going to be problematic with AoT or Webassembly. 

Not to address it now, however I'm wondering if we could precompute all these things at compile time via a source generator. The closure types will be known at compile time so we could potentially build a static map closure -> accessor.

Reply:
I was going to suggest that it might be beneficial to make this part of the razor compiler or a source generator or something, but really guys you are going so fast that I haven't had time to look into source generators yet :-)

Reply:
I've tested it with webassembly though and that seems to run fine on the IL interpreter. It also already contains the new linker hints to retain the type info for public fields for the closure types.

Reply:
I'll have a look at source generators now that I know that the razor compiler runs as one.